resolwe/flow/serializers/data.py

```diff
@@ -97,5 +97,12 @@ def validate_process(self, process):
     def validate_collection(self, collection):
         """Verify that changing collection is done in the right place."""
         if self.instance:
-            self.instance.validate_change_collection()
+            self.instance.validate_change_collection(collection)
         return collection
+
+    def update(self, instance, validated_data):
+        """Update."""
+        instance = super().update(instance, validated_data)
+        instance.move_to_collection(instance.collection)
```

Comment:
This triggers a save each time it is called, so you should put it in an if:
```
if instance.collection != validated_data['collection']:
```
But to be able to do that, you will have to copy a few parts of the original `update` method instead of calling `super`.

Reply:
Why calling original `update`? Doesn't `validated_data` hold the previous collection?

Reply:
No, `validated_data` holds the new state and `super` returns already updated object. So you have to get the access to [`instance`](https://github.com/encode/django-rest-framework/blob/aed74961ba03e3e6f53c468353f4e255eb788555/rest_framework/mixins.py#L65) in the original method. 
Not calling super will also save one save and thus improve performance.

Reply:
Oh, ok. Why not then just put pre-update collection in a variable `[pre_update_collection = instance.collection`](https://github.com/genialis/resolwe/pull/711/files#diff-85dabc7dcb33a71b4346421a5d041b48R105-R108).

Reply:
Where will you get the `instance`? You have to duplicate at least one line from the original function, so it is nicer if you copy all five lines as it is that short.

Reply:
Ok, I haven't noticed that it is passed in as an argument :)